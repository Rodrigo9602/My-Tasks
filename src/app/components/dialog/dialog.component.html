<div class="modal-wrapper">
    <div class="titleSelection" [ngSwitch]="data.form">
        <h2 mat-dialog-title *ngSwitchCase="'addTask'">Another task to the list...</h2>
        <h2 mat-dialog-title *ngSwitchCase="'updTask'">Lets change something here...</h2>        
    </div>


    <mat-dialog-content [ngSwitch]="data.form">
        <!--Add task form-->
        <form #customForm="ngForm" class="custom-form" id="add-form" *ngSwitchCase="'addTask'">
            <div class="input-wrapper">
                <input type="text" class="input-alt" placeholder="name" name="taskName"
                    pattern="^[a-zA-Z0-90-9áéíóúÁÉÍÓÚñÑ\s]+$" #taskName="ngModel"
                    [style.border]="!taskName.valid && taskName.touched ? '1px solid red':'1px solid var(--text-color)' "
                    [(ngModel)]="task.name" required>
                <div class="input-icon-alt">
                    <fa-icon [icon]="taskIcon" />
                </div>
            </div>

            <div class="input-wrapper">
                <textarea class="input-text" placeholder="description" name="taskDesc" #taskDesc="ngModel"
                    [(ngModel)]="task.description" required></textarea>
            </div>

            <div class="input-wrapper">
                <input type="text" class="input-alt date-input" placeholder="ending date" [min]="minDate" name="taskEnd" id="date"
                        #taskEnd="ngModel" [(ngModel)]="date" onfocus="(this.type='date')" onblur="(this.type='text')"
                        (ngModelChange)="onDateChange($event)" required>                

                
                <div class="input-icon-alt">
                    <fa-icon [icon]="dateIcon" />
                </div>
            </div>

            <mat-dialog-actions>
                <button class="btn btn-success" [mat-dialog-close]="data.dataObject" (click)="onSubmit()" [disabled]="customForm.invalid">Submit</button>
                <button type="button" class="btn btn-cancel" (click)="onCancel()">Cancel</button>
            </mat-dialog-actions>
        </form>
        <!--Update task form-->
        <div *ngIf="updateTaskData, else updateState">
            <form #customForm="ngForm" class="custom-form" id="upd-form" *ngSwitchCase="'updTask'">
                <div class="input-wrapper">
                    <input type="text" class="input-alt" placeholder="name" name="taskName"
                        pattern="^[a-zA-Z0-90-9áéíóúÁÉÍÓÚñÑ\s]+$" #taskName="ngModel"
                        [style.border]="!taskName.valid && taskName.touched ? '1px solid red':'1px solid var(--text-color)' "
                        [(ngModel)]="taskUpdate.name" required>
                    <div class="input-icon-alt">
                        <fa-icon [icon]="taskIcon" />
                    </div>
                </div>

                <div class="input-wrapper">
                    <textarea class="input-text" placeholder="description" name="taskDesc" #taskDesc="ngModel"
                        [(ngModel)]="taskUpdate.description" required></textarea>
                </div>

                <div class="input-wrapper">
                    <input type="text" class="input-alt date-input" placeholder="ending date" [min]="minDate" name="taskEnd"
                        #taskEnd="ngModel" [(ngModel)]="date" onfocus="(this.type='date')" onblur="(this.type='text')"
                        (ngModelChange)="onDateChange($event)" required>
                    <div class="input-icon-alt">
                        <fa-icon [icon]="dateIcon" />
                    </div>
                </div>

                <mat-select name="taskState" [(value)]="taskUpdate.state">
                    @for (state of states; track state) {
                    <mat-option value={{state}}>{{state}}</mat-option>
                    }
                </mat-select>

                <mat-dialog-actions>
                    <button class="btn btn-success" [mat-dialog-close]="data.dataObject" (click)="onSubmit()" [disabled]="customForm.invalid">Submit</button>
                    <button type="button" class="btn btn-cancel" (click)="onCancel()">Cancel</button>
                </mat-dialog-actions>
            </form>
        </div>

        <ng-template #updateState>
            <form #customForm="ngForm" class="custom-form" id="upd-form" *ngSwitchCase="'updTask'">
                <p>Only a change in the task status is allowed</p>
                <mat-select name="taskState" [(value)]="taskUpdate.state">
                    <mat-option value="en progreso" disabled>en progreso</mat-option>
                    <mat-option value="terminado">terminado</mat-option>
                    <mat-option value="no completado">no completado</mat-option>
                </mat-select>

                <mat-dialog-actions>
                    <button class="btn btn-success" [mat-dialog-close]="data.dataObject" (click)="onSubmit()" [disabled]="customForm.invalid">Submit</button>
                    <button type="button" class="btn btn-cancel" (click)="onCancel()">Cancel</button>
                </mat-dialog-actions>
            </form>
        </ng-template>       


    </mat-dialog-content>
   
</div>