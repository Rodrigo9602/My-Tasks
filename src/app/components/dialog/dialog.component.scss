.modal-wrapper {
    width: 100%;
    height: 600px;
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    justify-content: space-around;
    background-color: var(--bg);

    h2 {
        font-family: 'poppins';
        color: var(--text-color);
    }   
    .custom-form {
        width: 410px;        


        .date-input {
            width: 75%;
            padding-right: 20px;
        }
        
    }

    #upd-form {
        margin-top: -20px;
        mat-select {
            width: 80%;
            height: 2rem;
            border-radius: 0.5rem;
            padding: 5px 20px 0 20px;
            border: 1px solid var(--text-color);
            color: var(--text-color);
            margin: 0 auto;
        }
        
    }

    mat-dialog-actions {
        width: 100%;
        display: flex;        
        align-items: center;
        justify-content: center;
        gap: 30px;
    }

    
}