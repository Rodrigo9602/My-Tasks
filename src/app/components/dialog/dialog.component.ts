import { Component, Inject, OnInit } from '@angular/core';
import { MatDialogModule } from '@angular/material/dialog';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material/dialog';
import { MatSelectModule } from '@angular/material/select';
import { CommonModule, DatePipe } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { FontAwesomeModule } from '@fortawesome/angular-fontawesome';



import { TaskInterface } from '../../interfaces/task.interface';

import { faUser, faFileLines, faFileText, faCalendarTimes, faCalendar } from '@fortawesome/free-regular-svg-icons';

import { Task } from '../../models/task.model';



@Component({
  selector: 'app-dialog',
  standalone: true,
  imports: [CommonModule, MatDialogModule, FormsModule, MatSelectModule, FontAwesomeModule],
  providers: [DatePipe],
  templateUrl: './dialog.component.html',
  styleUrl: './dialog.component.scss'
})
export class DialogComponent implements OnInit {
  public task: TaskInterface;
  public taskUpdate: Task;
  public updateTaskData: boolean = false;
  private DateChanged: boolean = false;
  public date: any;
  public minDate: string = '';
  public states: Array<string> = ['creado', 'en progreso', 'terminado', 'no completado'];

  

  public nameIcon = faUser;
  public taskIcon = faFileLines;
  public descIcon = faFileText;
  public calendarIcon = faCalendar;
  public dateIcon = faCalendarTimes;

  constructor(
    public datePipe: DatePipe,
    public dialogRef: MatDialogRef<DialogComponent>,
    @Inject(MAT_DIALOG_DATA) public data: any,
  ) {
    this.task = {
      name: '',
      description: '',
      endingDate: new Date(),
    };

    this.taskUpdate = {
      _id: '',
      name: '',
      description: '',
      creationDate: new Date(),
      endingDate: new Date(),
      lastChange: new Date(),
      state: ''
    };
    

    const today = new Date();
    this.minDate = today.toISOString().split('T')[0];

  }

  ngOnInit(): void {
    if (this.data.form === 'updTask' && this.data.dataObject.task.state === 'creado') {
      this.updateTaskData = true;
      this.taskUpdate = JSON.parse(JSON.stringify(this.data.dataObject.task));
      this.formatDate();
    }
  }  




  formatDate() {
    const date = new Date(this.data.dataObject.task.endingDate);
    const localDate = new Date(date.getTime() + date.getTimezoneOffset() * 60 * 1000);
    this.date = this.datePipe.transform(localDate, 'dd/MM/yyyy');
  }

  onDateChange(event: any) {
    this.DateChanged = true;
    this.taskUpdate.endingDate = new Date(event);
  }
  

  onSubmit() {
    switch (this.data.form) {
      case 'addTask':
        this.task.endingDate = this.date;
        this.data.dataObject = this.task;
        this.dialogRef.close(this.data.dataObject);
        break;

      case 'updTask':
        if (this.data.dataObject.task.state === 'creado') {
          if (!this.DateChanged) {
            const utcDate = new Date(Date.UTC(
              this.date.split('/')[2],
              this.date.split('/')[1] - 1,
              this.date.split('/')[0]
            ));
            const offset = new Date().getTimezoneOffset();
            const date = new Date(utcDate.getTime() + offset * 60 * 1000);
            this.taskUpdate.endingDate = date;
          }
          this.data.dataObject.task = this.taskUpdate;
          this.data.dataObject.task.endingDate = this.taskUpdate.endingDate.toISOString();
          this.dialogRef.close(this.data.dataObject);
        } else {
          this.data.dataObject.task.state = this.taskUpdate.state;
          this.dialogRef.close(this.data.dataObject);
        }


        break;
    }
  }

  onCancel() {
    this.dialogRef.close();
  }
}
