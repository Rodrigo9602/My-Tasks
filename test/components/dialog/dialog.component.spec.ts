import { ComponentFixture, TestBed } from '@angular/core/testing';

import { DialogComponent } from '../../../src/app/components/dialog/dialog.component';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material/dialog';


describe('DialogComponent', () => {
  let component: DialogComponent;
  let fixture: ComponentFixture<DialogComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [DialogComponent],   
      providers: [
        { provide: MatDialogRef, useValue: {} },
        { provide: MAT_DIALOG_DATA, useValue: {} }
      ]
     
    })
    .compileComponents();
    
    fixture = TestBed.createComponent(DialogComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match initial snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });
});
