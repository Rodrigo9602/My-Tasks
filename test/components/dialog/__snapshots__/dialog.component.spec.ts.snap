// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialogComponent should match initial snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="modal-wrapper"
  >
    <div
      class="titleSelection"
    >
      
      
    </div>
    <mat-dialog-content
      class="mat-mdc-dialog-content mdc-dialog__content"
    >
      
      
      
      
    </mat-dialog-content>
  </div>
</div>
`;
